特定の講座への受講生たちのログイン率を返すAPIのDB処理のロジックを実装(DAIKI)

i-k-i-a-d opened the pull request:
・特定の講座への受講生たちのログイン率を返すAPIのDB処理のロジックを実装。
・中間テーブルであるattendancesテーブルを使ってリレーションメソッドをCourseモデルとStudentモデルの中に作成。
・ログイン率を返す処理までのルーティングとメソッドを作成。
・便宜上、AttendanceSeeder.phpファイルにテストデータを追加。

レビューよろしくお願いいたします。

i-k-i-a-d committed:
・特定の講座への受講生たちのログイン率を返すAPIのDB処理のロジックを実装。
・中間テーブルであるattendancesテーブルを使ってリレーションメソッドをCourseモデルとStudentモデルの中に作成。
・ログイン率を返す処理までのルーティングとメソッドを作成。
・便宜上、AttendanceSeeder.phpファイルにテストデータを追加。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +87,89 @@ public function show(AttendanceShowRequest $request) {
+     * @param int $courseId 
+     * @return boolean
+     */
+    public function idCheck ($courseId) {
```
真偽値を返す場合のメソッド名は、こちらの記事を参考に命名してみてください！
https://qiita.com/yskszk/items/5a7f99c974773f03a82a

kou-tech replied:
あとは、こちらの関数による講座の存在チェックの実装だと、講座データが多くなった場合にボトルネックとなる可能性があるので、全件取得 → ループしてチェックは避けた方が良いです！

i-k-i-a-d replied:
今回は、course_idが存在するかどうかのチェックは別タスクのバリデーション作成の方で実装予定なので、メソッド自体を削除しました。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +87,89 @@ public function show(AttendanceShowRequest $request) {
+        $monthAgo = $endDate->subMonth(1);
+        $yearAgo = $endDate->subYear(1);
+
+        if ($this->idCheck($courseId)) {
+            $course = Course::find($courseId);
+            $studentsCount = $course->students()->get()->count();
+
+            foreach ($course->students()->get() as $student) {
+                if ($student->last_login_at >= $weekAgo) {
+                    $weekCount++; 
+                } 
+                if ($student->last_login_at >= $monthAgo) {
+                    $monthCount++;
+                } 
+                if ($student->last_login_at >= $yearAgo) {
+                    $yearCount++;
+                }   
+            }
+    
+            $last_week_login_rate = $this->num2per($weekCount, $studentsCount);
+            $last_month_login_rate = $this->num2per($monthCount, $studentsCount);
+            $last_year_login_rate = $this->num2per($yearCount, $studentsCount);
+    
+            return response()->json([
+                'last_week_login_rate' => $last_week_login_rate, 
+                'last_month_login_rate' => $last_month_login_rate,
+                'last_year_login_rate' => $last_year_login_rate,
+            ], 200);
```
if分のネストは、可読性を悪くします。
そのため、早期returnなどで解消ができる場合があるので、調べて書き直してみてください！

i-k-i-a-d replied:
course_idのチェックをController内で行わないことにしたので、ネストの問題も解消しました。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +87,89 @@ public function show(AttendanceShowRequest $request) {
+     * @return json
+     */ 
+    public function loginRate($courseId) {
+        $weekCount = 0;
+        $monthCount = 0;
+        $yearCount = 0;
+        $endDate = now(); 
+        $weekAgo = $endDate->subWeek(1);
+        $monthAgo = $endDate->subMonth(1);
+        $yearAgo = $endDate->subYear(1);
```
こちらは、一度に全ての周期のログイン率を返すのではなく、リクエストの周期種別に合わせてログイン率を返すようにしたいです。例えば、weekを指定したら、1週間のログイン率を返すようなイメージです！

i-k-i-a-d replied:
7日、30日、365日の周期で、それぞれ異なるURLでアクセスし、レスポンスを返せるように実装しております。

i-k-i-a-d replied:
実際のページではまとめて表示するにも関わらず、別々のURLでアクセスして情報を3つ個別に取得するメリットをよかったら教えてほしいです。

kou-tech commented on app/Model/Course.php:
```diff
@@ -67,4 +67,9 @@ protected static function boot()
         });
     }
+
+    //受講生の情報を取得。
+    public function students () {
+        return $this->belongsToMany(Student::class, 'attendances', 'course_id', 'student_id');
+    }
```
生徒と講座は受講テーブルを通しての、多対多とも捉えれるのですが、現状はいくつもリレーション定義することは実装が複雑になるので、現状のリレーション関係だけで対応いただきたいです！

i-k-i-a-d replied:
今回はloginRateメソッドを多対多のリレーションメソッドを使わずに実装したので、これは削除しました。

kou-tech commented on app/Model/Student.php:
```diff
@@ -112,4 +112,9 @@ public function getSexAttribute($value)
         return null;
     }
+
+    //生徒が受講している講座情報を取得。
+    public function manyCourses () {
+        return $this->belongsToMany(Course::class, 'attendances', 'student_id', 'course_id');
+    }
```
生徒と講座は受講テーブルを通しての、多対多とも捉えれるのですが、現状はいくつもリレーション定義することは実装が複雑になるので、現状のリレーション関係だけで対応いただきたいです！

i-k-i-a-d replied:
今回はloginRateメソッドを多対多のリレーションメソッドを使わずに実装したので、これは削除しました。

kou-tech commented on database/seeds/AttendanceSeeder.php:
Attendanceだけを追加してしまうと、データの不整合が発生してしまいます。
Attendanceを追加する場合は、database/seeds/LessonAttendanceSeeder.phpも追加する必要があります！

i-k-i-a-d replied:
AttendancesテーブルとLessonAttendancesテーブルの関係を見落としてました。
for文でまとめてダミーデータを作成するように追記してます。

i-k-i-a-d committed:
loginRateメソッドをリレーションメソッドを使わない方法で修正。
CourseモデルとStudentモデルからリレーションメソッドを削除。
loginRateメソッド内でif分の
ネストが起きないように修正。
loginRateメソッドのレスポンスを周期ごとに別々に返せるように実装。
course_idが存在するかどうかのチェックは別タスクのバリデーションに任せるので、idチェック用のメソッドは削除。
loginRateメソッドで周期を受け取れるようにルーティングのルートパラメータに{period}を追加。
AttendanceSeederにダミーデータを追加したので、整合性を整えるため、LessonAttendanceSeederにもダミーデータを追加。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -5,6 +5,8 @@
 use App\Model\Attendance;
 use App\Model\Lesson;
+use App\Model\Student;
+use App\Model\Course;
```
こちらは参照していないので、削除お願いします。

````php
use App\Model\Course;
````

i-k-i-a-d replied:
削除しました。
前任者がこれを引数を受取る$courseIdの前に記述してましたが、これってなんか意味があったんでしょうか？

kou-tech replied:
ルートモデルバインディングを利用したようですが、何かと不評を聞くので利用しなくて良いです！

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +88,69 @@ public function show(AttendanceShowRequest $request) {
+            }   
+        }
+        
+        if ($period === "7") {
+            $last_week_login_rate = $this->loginRateCalculate($weekCount, $studentsCount);
+            return response()->json(['last_week_login_rate' => $last_week_login_rate,], 200);
+        } elseif ($period === "30") {
+            $last_month_login_rate = $this->loginRateCalculate($monthCount, $studentsCount);
+            return response()->json(['last_month_login_rate' => $last_month_login_rate,], 200);
+        } elseif ($period === "365") {
+            $last_year_login_rate = $this->loginRateCalculate($yearCount, $studentsCount);
+            return response()->json(['last_year_login_rate' => $last_year_login_rate,], 200);
+        } else {
+            return response()->json([
+                'result' => false, 
+                'error_message' => 'Invalid Request Body.',
+            ], 400);
+        }
```
periodは、種別値で次の仕様でした。
お手数ですが、次の条件式で書き直しをお願いします。
`week`、`month`、`year`

kou-tech replied:
elseは不要なので、次のようにお願いします。

````php
        } else if {

        }

        return response()->json([
            'result' => false,
            'error_message' => 'Invalid Request Body.',
        ], 400);
````

i-k-i-a-d replied:
周期の記述に関しては修正しました。

しかし、elseに関しては、今回loginRateメソッドのコードをより簡潔にリファクタリングするために必要だったので、記述したままにしてます。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +88,69 @@ public function show(AttendanceShowRequest $request) {
+        $endDate = now();  
+        $weekAgo = $endDate->subWeek(1);
+        $monthAgo = $endDate->subMonth(1);
+        $yearAgo = $endDate->subYear(1);
```
今回は、リクエストから来た周期の種別に対して、日付の差分を取得できればいいので、こちらの記述は冗長な気がしました。
1週間分だけほしい場合は、$monthCountとかは不要ですよね。その辺りを最適化できるのではと思いました！

i-k-i-a-d replied:
一つの変数にまとめて条件分岐で中身を切り替えるようにリファクタリングしてます。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +88,69 @@ public function show(AttendanceShowRequest $request) {
+        $students = Attendance::where('course_id', $courseId)->get();
+        $studentsCount = Attendance::where('course_id', $courseId)->get()->count();
+
+        foreach ($students as $student) {
+            $lastLoginDate = Student::findOrFail($student->id)->last_login_at;
+            if ($lastLoginDate >= $weekAgo) {
+                $weekCount++;
+            } 
+            if ($lastLoginDate >= $monthAgo) {
+                $monthCount++;
+            } 
+            if ($lastLoginDate >= $yearAgo) {
+                $yearCount++;
+            }   
+        }
```
上記の修正に合わせて、この辺りもより最適化できそうです！

i-k-i-a-d replied:
ここも同じくリファクタリング済みです。

kou-tech commented on app/Model/Course.php:
```diff
@@ -67,4 +67,5 @@ protected static function boot()
             }
         });
     }
+
 }
```
完全に修正以前の状態に戻す場合、ここでファイルの差分が検出されないはずなので、もう一度ご確認お願いします！

i-k-i-a-d replied:
リモートリポジトリがそのような仕様になっていたとは初耳でした。
余分な空行は消しました。

kou-tech commented on app/Model/Student.php:
```diff
@@ -112,4 +112,5 @@ public function getSexAttribute($value)
 
         return null;
     }
+
 }
```
完全に修正以前の状態に戻す場合、ここでファイルの差分が検出されないはずなので、もう一度ご確認お願いします！

i-k-i-a-d replied:
余分な空行は消しました。

kou-tech commented on database/seeds/LessonAttendanceSeeder.php:
```diff
@@ -57,5 +57,20 @@ public function run()
+                    ],
+                ]);
+            }
+        }
```
受講テーブルに追加した講座に対してのレッスンidの数だけ、慎重に調整する必要がありました。
例えばですが、講座ID = 4 のレッスンはないので、1つもレコードはないはずです。
不整合のデータがあると、フロントエンドの実装に影響があるので、もう一度ご確認お願いします！

i-k-i-a-d replied:
今回の失敗の要因はテーブル同士の関連性に無頓着だったことです。
そもそも、現状生徒が受講できるlessonが振られている講座は講座１のみなので、講座２以上の講座を受講しているダミーデータをattendancesテーブルに入れようとすること自体ナンセンスでした。
新たにlessonを割り振るという方法もあったんですが、他の人のタスクに影響する可能性もあるため、今回は控えました。

kou-tech reviewed (commented):
レビューしました！

> 別々のURLでアクセスして情報を3つ個別に取得するメリット

前提として、どちらも１長１短のメリットとデメリットがあります。
この設計のメリット
柔軟性がまず向上します。クライアントは必要に応じて特定の周期のデータだけを取得することができるため、もし画面の変更があっても不要なデータを取得する必要がないです。

スケーラビリティが良くなります。
将来的に新しい周期（例: 6ヶ月ごと、3年ごとなど）を追加したい場合、新しいAPIエンドポイントを簡単に追加することができます。

i-k-i-a-d committed:
loginRateメソッドの処理でまとめられる部分をリファクタリング実施。
CourseモデルとStudentモデルの余計な空行を削除。
attendancesテーブルとlesson_attendancesテーブル内のダミーデータの整合性がとれるように、それぞれSeederファイルを修正。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +88,63 @@ public function show(AttendanceShowRequest $request) {
+        } elseif ($period === "year") {
+            $periodAgo = $endDate->subYear(1);
+            $loginRateKey = 'last_year_login_rate';
+        } else {
+            return response()->json([
+                'result' => false, 
+                'error_message' => 'Invalid Request Body.',
+            ], 400);
+        }   
```
elseは$periodのバリデーションに委託すれば、不要なので削除しましょうか。

````php
        } else {
            return response()->json([
                'result' => false, 
                'error_message' => 'Invalid Request Body.',
            ], 400);
        }   
````

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +88,63 @@ public function show(AttendanceShowRequest $request) {
+        }
+
+        $loginRate = $this->loginRateCalculate($periodCount, $studentsCount);
+        return response()->json([$loginRateKey => $loginRate,], 200);
```
keyは`login_rate`で問題ないです！

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +88,63 @@ public function show(AttendanceShowRequest $request) {
+        
+    }
+
+    /**
+     * %計算
+     * @param int $number, $total, $precision 
+     * @return int
+     */
+    public function loginRateCalculate($number, $total, $precision = 0) {
+        if ($number < 0) {
+          return 0;
+        }
+      
+        try {
+            $percent = ($number / $total) * 100;
+            return round($percent, $precision);
+        } catch (\Throwable $e) {
+            return 0;
+        }
+    }
```
パーセント計算は、コンテキストによって計算方法が変わりやすいので、一旦は切り捨てで問題ないです。
あとは、phpdocの書き方が誤っています。その他、関数名に適したコメントを残すのと、好みにもよりますが関数は動詞が先に来ることが多いです。それを踏まえると次のようなコードになるかと思います。

````php
    /**
     * 受講生ログイン率計算
     *
     * @param int $number
     * @param int $total
     * @return int
     */
    public function calcLoginRate($number, $total) {
        if ($total === 0) return 0;
        return round($number / $total * 100);
    }
````

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +88,63 @@ public function show(AttendanceShowRequest $request) {
+        $studentsCount = $students->count();
+
+        foreach ($students as $student) {
+            $lastLoginDate = Student::findOrFail($student->id)->last_login_at;
```
Eager Loadingを考慮して、受講テーブルを取得するときに生徒の情報も取得するようにしましょう。
データベースのI/Oを減らすことができます。

i-k-i-a-d committed:
引数で受け取った文字が不正だった場合の処理は次回タスクのバリデーションに任せるので今回はelseでの条件分岐は削除。
で受け取った値に応じてキーの文字列を変えていたが、login_rateで固定した。それに伴い不要なコードを削除。
ログイン率の計算方法を四捨五入ではなく、切り捨てに変更。
ログイン率を計算する関数名を変更。
ログイン率を計算する関数のphpDocを修正。
withメソッドを使って、DBへの接続回数を減らせるように修正。
ルートモデルバインディングとは何かわからなかったが、ロジックに不要となったuseを削除。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +86,55 @@ public function show(AttendanceShowRequest $request) {
+          return 0;
+        }
+      
+        try {
+            $percent = ($number / $total) * 100;
+            return floor($percent);
+        } catch (\Throwable $e) {
+            return 0;
+        }
```
number<0になるのは確かにおかしいですが、この書き方だと問題の発生に気づかないため、書かない方が良いと思いました。また、現状のコードで起こりうる可能性は少ないため、一度削除しましょう。
起こりうることを事前に防ぐ場合は、例外に投げて適切なログ出力等が必要になります。(今回のタスクには服なくて良いです。)

````php
        if ($total < 0) {
          return 0;
        }
````

kou-tech replied:
try~catchはどういった意図で書かれているかを教えていただきたいです！

i-k-i-a-d replied:
条件分岐は削除しました。
また、try catchを書いた意図としては、$totalや$numberがもし0の値になった時にエラーがでるので、その時に0を返せるように記述しております。

kou-tech replied:
$numbre=0の場合は、0で問題ないですよね。そして、0徐算は私が提示した以前のコードで対応できませんか？

````php
if ($total === 0) return 0;
````

i-k-i-a-d replied:
try catchを消して修正しました。_

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +86,55 @@ public function show(AttendanceShowRequest $request) {
+     * @return json
+     */ 
+    public function loginRate($courseId, $period) {
+        $periodCount = 0;
```
現場にもよりますが、一般的には変数のスコープはなるべく小さくしたいので、使用する直前に定義しましょう！

````php
        // 期間内にログインした受講生数
        $periodCount = 0;
        foreach ($students as $student) {
            $lastLoginDate = $student->student->last_login_at;
            if ($lastLoginDate >= $periodAgo) {
                $periodCount++;
            }
        }
````

i-k-i-a-d replied:
変数定義の位置を修正しました。
スコープをなるべく小さくするのはデバッグがやりやすいようにでしょうか？

kou-tech replied:
一般的には使用範囲が狭くなることで、コードの可読性が良くなります。
場合によっては、変数が使いまわされる可能性もあるので、小さくすることで、バグの発生を防ぐことにもつながります。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +86,55 @@ public function show(AttendanceShowRequest $request) {
+     */ 
+    public function loginRate($courseId, $period) {
+        $periodCount = 0;
+        $endDate = now();  
```
こちらはLaravelのヘルパー関数なのですが、実質次のコードと同じです。
ヘルパー関数は一見便利なのですが、テスタビリティが悪かったり、どういった値が返ってくるか分かりづらいので次のように明示的にするのが無難です。

````php
        $endDate = new Carbon();
````

i-k-i-a-d replied:
Carbonクラスをインスタンス化する方法に修正しました。

ただ「ヘルパー関数は一見便利なのですが、テスタビリティが悪かったり、どういった値が返ってくるか分かりづらい」という部分がいまいち理解できなかったので、具体例などと一緒に解説いただけるとありがたいです。

kou-tech replied:
テストコードはまだわからないと思いますが、テストするにあたって依存しているクラス等はモックという擬似クラスに置き換わることがあります。しかし、ヘルパー関数がそれがやりづらいため、テストするにあたって使いづらいコードとなってしまいます。

どういった値が返ってくるかに関しては、`now()`を使うことで、`Carbon`のインスタンスがこれで返ってくることが予想しづらいといった感じです。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +86,55 @@ public function show(AttendanceShowRequest $request) {
+        $studentsCount = $students->count();
+
+        foreach ($students as $student) {
+            $lastLoginDate = $student->student->last_login_at;
+            if ($lastLoginDate >= $periodAgo) {
+                $periodCount++;
+            } 
```
$lastLoginDateも、$student->student->last_login_atもCabonのインスタンスなので、専用の日付比較メソッドがあります。日付の比較は何かとバグにつながりやすいので、ライブラリに従って次のように修正をお願いします。

````php
            $lastLoginDate = $student->student->last_login_at;
            if ($lastLoginDate->greaterThanOrEqualTo($periodAgo)) {
                $periodCount++;
            }
````

i-k-i-a-d replied:
greaterThanOrEqualToを知らなかったので、調べたところ、gte()と一緒だと知ったので、短い方を採用しました。

また、$lastLoginDateはAttendanceモデル内のリレーションメソッドを使って、studentsテーブルから取得した値なんですが、どうしてCarbonのインスタンスになるのか解説いただけると幸いです。

kou-tech replied:
Studentのモデルで特定のプロパティはキャストによって型変換されているからです。
dateがLaravelでは、Carbonインスタンスに該当します。

````Student.php
    /**
     * キャスト
     */
    protected $casts = [
        'birth_date' => 'date',
        'last_login_at' => 'date',
    ];
````

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +86,55 @@ public function show(AttendanceShowRequest $request) {
+            $periodAgo = $endDate->subYear(1);
+        } 
+
+        $students = Attendance::with('student')->where('course_id', $courseId)->get();
+        $studentsCount = $students->count();
```
こちら正確には、生徒ではなく受講状況のレコードのコレクションを取得するので、適した変数名を命名しましょう。
今後、他の実装者や未来の自分でも理解できるコードをなるべく書くことを意識しましょう。

i-k-i-a-d replied:
実質、特定の講座を受講している生徒の情報を取得するための記述だったので、こっちの方がわかりやすいかと思ってました。

kou-tech replied:
確かに捉え方次第ではありますが、今回は修正いただいた`$attendances`の方が文脈的には正しいと思います。
元にその後のループ処理で`$students->student`のコードに違和感を覚えると思います。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +86,55 @@ public function show(AttendanceShowRequest $request) {
         ]);
     }
+
+     /**
+     * 受講生ログイン率取得API
+     * @param string courseId
+     * @return json
+     */ 
```
phpdocの書き方は次のようにお願いします。
リクエストファイルや、リソースファイルが実装されたらまた更新が必要になります。

````php
    /**
     * 受講生ログイン率取得API
     *
     * @param string $courseId
     * @param string $period
     * @return \Illuminate\Http\JsonResponse
     */
````

i-k-i-a-d replied:
戻り値でjsonを返すときの記述方法を知らなかったため、あのような記述になってました。

kou-tech replied:
他の方の実装コードは全然参考にしていただいて問題ないです！

i-k-i-a-d committed:
ログイン率を計算するメソッド内の条件分岐を削除。
ログインしていた生徒の数をカウントする変数の定義位置を修正。
現在の日時を取得する方法をCarbonクラスをインスタンス化する方法に修正。
日付の比較を比較演算子ではなく、gte();メソッド利用するように修正。
受講状況のレコードのコレクションを入れる変数名を修正。
受講生ログイン率取得APIのphpDocを修正。

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +87,54 @@ public function show(AttendanceShowRequest $request) {
         ]);
     }
+
+     /**
+     * 受講生ログイン率取得API
+     * @param string $courseId
+     * @param string $period
+     * @return \Illuminate\Http\JsonResponse
+     */ 
```
次のように改行を追加して修正をお願いします。

````php
    /**
     * 受講生ログイン率取得API
     *
     * @param string $courseId
     * @param string $period
     * @return \Illuminate\Http\JsonResponse
     */
````

kou-tech commented on app/Http/Controllers/Api/Instructor/AttendanceController.php:
```diff
@@ -86,4 +87,54 @@ public function show(AttendanceShowRequest $request) {
+        return response()->json(['login_rate' => $loginRate], 200);
+    }
+
+    /**
+     * 受講生ログイン率計算
+     * @param int $number
+     * @param int $total
+     * @return int
+     */
```
次のように改行を追加して修正をお願いします。

````php
    /**
     * 受講生ログイン率取得API
     *
     * @param string $courseId
     * @param string $period
     * @return \Illuminate\Http\JsonResponse
     */
````

kou-tech reviewed (commented):
お手数おかけしますが、`database/seeds/LessonAttendancesSeeder.php`という似たような使われていないSeederがあるため、こちらを削除していただくようにお願いします！

i-k-i-a-d committed:
ログイン率を計算するメソッド内の0除算のロジックをtry catchではなく、if文で実装。
phpDocのコメントの後に改行を入れた。

i-k-i-a-d committed:
という変数名をわかりやすいようにに変更。
